cmd/prepare-storage.go

```diff
@@ -88,6 +88,87 @@ func formatXLCleanupTmpLocalEndpoints(endpoints EndpointList) error {
 	return nil
 }
 
+// validate reference format against list of XL formats.
+func validateXLFormats(format *formatXLV3, formats []*formatXLV3, endpoints EndpointList, setCount, drivesPerSet int) error {
+	for i := range formats {
+		if formats[i] == nil {
+			continue
+		}
+		if err := formatXLV3Check(format, formats[i]); err != nil {
+			return fmt.Errorf("%s format error: %s", endpoints[i], err)
+		}
+	}
+	if len(format.XL.Sets) != setCount {
+		return fmt.Errorf("Current backend format is inconsistent with input args (%s), Expected set count %d, got %d", endpoints, len(format.XL.Sets), setCount)
+	}
+	if len(format.XL.Sets[0]) != drivesPerSet {
+		return fmt.Errorf("Current backend format is inconsistent with input args (%s), Expected drive count per set %d, got %d", endpoints, len(format.XL.Sets[0]), drivesPerSet)
+	}
+	return nil
+}
+
+// connect to list of endpoints and load all XL disk formats, validate the formats are correct
+// and are in quorum, if no formats are found attempt to initialize all of them for the first
+// time. additionally make sure to close all the disks used in this attempt.
+func connectLoadInitFormats(firstDisk bool, endpoints EndpointList, setCount, drivesPerSet int) (*formatXLV3, error) {
+	storageDisks, err := initStorageDisks(endpoints)
+	if err != nil {
+		return nil, err
+	}
+	defer closeStorageDisks(storageDisks)
+
+	// Attempt to load all `format.json` from all disks.
+	formatConfigs, sErrs := loadFormatXLAll(storageDisks)
+
+	// Pre-emptively check if one of the formatted disks
+	// is invalid. This function returns success for the
+	// most part unless one of the formats is not consistent
+	// with expected XL format. For example if a user is
+	// trying to pool FS backend into an XL set.
+	if err = checkFormatXLValues(formatConfigs); err != nil {
+		return nil, err
+	}
+
+	for i, sErr := range sErrs {
+		if _, ok := formatCriticalErrors[errors.Cause(sErr)]; ok {
+			return nil, fmt.Errorf("Disk %s: %s", endpoints[i], sErr)
+		}
+	}
+
+	if shouldInitXLDisks(sErrs) {
+		if !firstDisk {
+			return nil, errNotFirstDisk
+		}
+		return initFormatXL(storageDisks, setCount, drivesPerSet)
+	}
+
+	// Following function is added to fix a regressions which was introduced
+	// in release RELEASE.2018-03-16T22-52-12Z after migrating v1 to v2 to v3.
+	// This migration failed to capture '.This' field properly which indicates
+	// the disk UUID association. Below function is called to handle and fix
+	// this regression, for more info refer https://github.com/minio/minio/issues/5667
+	if err = fixFormatXLV3(storageDisks, endpoints, formatConfigs); err != nil {
+		return nil, err
+	}
+
+	// If any of the .This field is still empty we wait on them to be fixed.
+	if formatXLV3ThisEmpty(formatConfigs) {
+		return nil, errInvalidArgument
```

Comment:
we can return a more specific error `errFormatThisEmpty` so that in case we return `errInvalidArgument` from other part of the code we dont ignore it.